Allow chargers to provide vehicle SoC via ISO15118

DerAndereAndi opened the pull request:
Some chargers (e.g. Porsche/Audi Mobile Charger Connect) can provide the EVs current SoC via the ISO15118 connection to the car.

This change allows chargers to implement the SoC method and return the SoC or api.ErrNotAvailable if the currently connected car does not provide the data.

As the SoC via this connection is always up to date, the estimator does not need to run any more

andig commented on core/soc/estimator.go:
```diff
@@ -91,6 +93,15 @@ func (s *Estimator) RemainingChargeEnergy(targetSoC int) float64 {
+	if charger, ok := s.charger.(api.Battery); ok {
+		f, err := charger.SoC()
+
+		if !errors.Is(err, api.ErrNotAvailable) {
```
eigentlich if err == nil, oder? Falls es irgendeinen Fehler gäbe bliebe ja immer noch das Fallback zum Fahrzeug Soc?

DerAndereAndi committed:
Only succeed if no error occurs

DerAndereAndi committed:
Allow chargers to provide vehicle SoC via ISO15118

Some chargers (e.g. Porsche/Audi Mobile Charger Connect) can provide the EVs current SoC via the ISO15118 connection to the car.

This change allows chargers to implement the SoC method and return the SoC or api.ErrNotAvailable if the currently connected car does not provide the data.

As the SoC via this connection is always up to date, the estimator does not need to run any more

andig commented on core/soc/estimator_test.go:
```diff
@@ -28,13 +29,14 @@ func TestRemainingChargeDuration(t *testing.T) {
 
 func TestSoCEstimation(t *testing.T) {
 	ctrl := gomock.NewController(t)
+	charger := mock.NewMockCharger(ctrl)
```
Brauchts den? Ohne Battery interface kannst du da auch nil rein reichen

DerAndereAndi replied:
Ne, habs entfernt

DerAndereAndi committed:
Simplify test

DerAndereAndi committed:
Add check if SoC is provided via the charger

andig commented:
@DerAndereAndi hast Du mal getestet ob Restzeit etc jetzt auch korrekt aktualisiert werden? Evtl. kleiner Testcase? Dem Augenschein nach sieht das gut aus!

DerAndereAndi committed:
Added test and fixed logic

andig commented on core/soc/estimator.go:
```diff
@@ -93,33 +93,40 @@ func (s *Estimator) RemainingChargeEnergy(targetSoC int) float64 {
+	var fetchedSoC float64
+	var err error
+
+	socChargerSucceeded := false
```
die kannst Du sparen wenn Du fetchedSoC als *float64 machst- das lässt sich dann auf nil checken. *bool ist mein Lieblingstristate Typ ;)

andig commented on core/soc/estimator_test.go:
```diff
@@ -28,14 +29,20 @@ func TestRemainingChargeDuration(t *testing.T) {
 }
 
 func TestSoCEstimation(t *testing.T) {
```
wären nicht 2 getrennte testcases einfacher?

DerAndereAndi replied:
Der code ist halt zu 99% identisch, finde ich so besser. Wenn man nen Test hinzufügt ist er gleich  für beide Szenarien abgedeckt

DerAndereAndi committed:
Simplify a bit

DerAndereAndi commented:
@andig Fehlt noch etwas?

andig commented:
@DerAndereAndi kannst Du hier bitte nochmal rein schauen: https://github.com/andig/evcc/blame/master/core/soc/estimator.go#L114

Da sind m.E. 2 Dinge fehlerhaft:
- wenn ein Fehler auftritt wird mit den Daten weiter gearbeitet, die sind aber nicht verlässlich
- der Fehler wird erst geloggt und dann zurück gegeben, so dass am Ende 2x geloggt wird, siehe https://github.com/andig/evcc/pull/1213#issuecomment-901624365 im Logfile

andig commented:
@DerAndereAndi kannst Du hier bitte nochmal rein schauen: https://github.com/andig/evcc/blame/master/core/soc/estimator.go#L114

Da sind m.E. 2 Dinge fehlerhaft:
- wenn ein Fehler auftritt wird mit den Daten weiter gearbeitet, die sind aber nicht verlässlich
- der Fehler wird erst geloggt und dann zurück gegeben, so dass am Ende 2x geloggt wird, siehe

DerAndereAndi commented:
@andig also in Zeile 114 ist der Vehicle SoC, nicht vom Charger. Den Code habe ich nicht geändert, siehe den Stand davor: https://github.com/andig/evcc/blob/4fbdb526b0362951c3f000e37a48cf0c74e40e48/core/soc/estimator.go#L100

Das hat nichts mit diesem PR zu tun, sondern ist ein Problem der originären Implementierung. Sollten wir vielleicht als extra Ticket machen.

andig commented:
Kann gut sein :/ PR wäre toll :)
